cpp/include/raft/comms/mpi_comms.hpp

```diff
@@ -0,0 +1,290 @@
+#include <raft/handle.hpp>
+
+#define MPI_TRY(call)                                                          \
+  do {                                                                         \
+    int status = call;                                                         \
+    if (MPI_SUCCESS != status) {                                               \
+      int mpi_error_string_lenght = 0;                                         \
+      char mpi_error_string[MPI_MAX_ERROR_STRING];                             \
+      MPI_Error_string(status, mpi_error_string, &mpi_error_string_lenght);    \
+      RAFT_EXPECTS(MPI_SUCCESS == status, "ERROR: MPI call='%s'. Reason:%s\n", \
+                   #call, mpi_error_string);                                   \
+    }                                                                          \
+  } while (0)
+
+#define MPI_TRY_NO_THROW(call)                                              \
+  do {                                                                      \
+    int status = call;                                                      \
+    if (MPI_SUCCESS != status) {                                            \
+      int mpi_error_string_lenght = 0;                                      \
+      char mpi_error_string[MPI_MAX_ERROR_STRING];                          \
+      MPI_Error_string(status, mpi_error_string, &mpi_error_string_lenght); \
+      printf("MPI call='%s' at file=%s line=%d failed with %s ", #call,     \
+             __FILE__, __LINE__, mpi_error_string);                         \
+    }                                                                       \
+  } while (0)
+
+namespace raft {
+namespace comms {
+
+static MPI_Datatype get_mpi_datatype(const datatype_t datatype) {
+  switch (datatype) {
+    case datatype_t::CHAR:
+      return MPI_CHAR;
+    case datatype_t::UINT8:
+      return MPI_UNSIGNED_CHAR;
+    case datatype_t::INT32:
+      return MPI_INT;
+    case datatype_t::UINT32:
+      return MPI_UNSIGNED;
+    case datatype_t::INT64:
+      return MPI_LONG_LONG;
+    case datatype_t::UINT64:
+      return MPI_UNSIGNED_LONG_LONG;
+    case datatype_t::FLOAT32:
+      return MPI_FLOAT;
+    case datatype_t::FLOAT64:
+      return MPI_DOUBLE;
+    default:
+      // Execution should never reach here. This takes care of compiler warning.
+      return MPI_DOUBLE;
+  }
+}
+
+static MPI_Op get_mpi_op(const op_t op) {
+  switch (op) {
+    case op_t::SUM:
+      return MPI_SUM;
+    case op_t::PROD:
+      return MPI_PROD;
+    case op_t::MIN:
+      return MPI_MIN;
+    case op_t::MAX:
+      return MPI_MAX;
+    default:
+      // Execution should never reach here. This takes care of compiler warning.
+      return MPI_MAX;
+  }
+}
+
+class mpi_comms : public comms_iface {
+ public:
+  mpi_comms(MPI_Comm comm, const bool owns_mpi_comm)
+    : _owns_mpi_comm(owns_mpi_comm),
+      _mpi_comm(comm),
+      _size(0),
+      _rank(1),
+      _next_request_id(0) {
```

Comment:
I think we agreed to use trailing _ (instead of preceding _) for class private variables.

https://github.com/rapidsai/raft/blob/branch-0.16/cpp/.clang-tidy#L185

And see

https://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier

using preceding _ is generally discouraged.